armi/reactor/blueprints/__init__.py

```diff
@@ -326,7 +326,8 @@ def _prepConstruction(self, cs):
                     for a in list(self.assemblies.values())
                     if not any(a.hasFlags(f) for f in assemsToSkip)
                 )
-                axialExpansionChanger.axialExpansionChanger.expandColdDimsToHot(
+                chngrT = getPluginManager().hook.getAxialExpansionChanger()
+                chngrT.expandColdDimsToHot(
```

Comment:
Okay, so the `pluggy` docs don't provide any particular tools for failing if no plugin implements a hook. So we are clearly supposed to handle that ourselves.

I guess I would say if `chngrT` (strange variable name) is None, the code shouldn't fail, or throw a warning. I would prefer it if this line was wrapped in an `if`, and the model was allowed to continue without failure.

Or would this cause a failure elsewhere in the run?

Reply:
Yeah the docs don't have anything on that case. But a test file shows if nothing implements the hook, in this case we just get `None` returned

<details>

<summary>Demo no hook impl</summary>

```python
"""Play around with plugins"""

import pluggy

spec = pluggy.HookspecMarker("demo")
impl = pluggy.HookimplMarker("demo")

class MySpec:
    @staticmethod
    @spec(firstresult=True)
    def get_number() -> float:
        """Give me a number, any number."""

pm = pluggy.PluginManager("demo")
pm.add_hookspecs(MySpec)
result = pm.hook.get_number()
assert result is None, result
```
</details>

I'll add a condition if we get `None` in the event an application that inherits from `armi.apps.App` un-registers the `ReactorPlugin` and also does not provide their own implementation of this hook.

I do not expect this to cause any later failures. This would just result in not doing any axial expansion when construction assemblies from blueprints.

Reply:
86be86c7

Reply:
Sorry, I may not be clear.

I don't want people who have reactors without assemblies to HAVE to implement this hook.  That is a waste of their time.

So, I would prefer if people who were modeling liquid-fuel reactors (for instance) to not have to understand this took, and they can just skip it.

So, all we have to do here is something like:

```python
                chngrT = getPluginManager().hook.getAxialExpansionChanger()
                if chngrT is not None:
                    chngrT.expandColdDimsToHot(
```

Reply:
I'm sorry if I haven't been clear.  But since not all reactors _have_ assemblies that would apply to this hook, I expect to in the near future the default `ReactorPlugin` probably won't implement this hook. That will be something we move elsewhere, to make this feature invisibly optional.

This is another case of ARMI assuming that everyone is modeling "pin-type reactors with hex-shaped assemblies".  The default functionality of ARMI should not be so specific, though it was in many places, and still is. So we have been making a concerted effort to remove such assumptions.

Imagine in the future that ARMI has a default `App`, but also a `HexAssemlyApp`. And the `HexAssemblyApp` pulls in a ton of details like this one. It might also pull in a new `HexAssemblyCore` base class for people to subclass.

Maybe the real solution won't look like that exactly, but I have heard that ARMI's many assumptions about reactor type causes non-TerraPower users a lot of wasted effort. So we are working on that. (Actually, I heard that from _you_ Drew. You are one of the people who told me that. I would love to pick your brain about what ARMI's assumptions wasted your team's time in the past.)

Reply:
The changes you requested were added in 86be86c7c01f272ff78d02f334e22f18a6b59f40

https://github.com/terrapower/armi/blob/86be86c7c01f272ff78d02f334e22f18a6b59f40/armi/reactor/blueprints/__init__.py#L329-L334

I support these changes for the reasons you laid out: fewer assumptions about the reactor someone is modeling the better.